src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/ActivityExtensions.cs

```diff
@@ -18,41 +19,51 @@ namespace OpenTelemetry.Exporter.OpenTelemetryProtocol.Implementation;
 internal static class ActivityExtensions
 {
     private static readonly ConcurrentBag<ScopeSpans> SpanListPool = new();
+    private static Dictionary<string, ScopeSpans>? spansByLibraryCached;
 
     internal static void AddBatch(
         this ExportTraceServiceRequest request,
         SdkLimitOptions sdkLimitOptions,
         Resource processResource,
         in Batch<Activity> activityBatch)
     {
-        Dictionary<string, ScopeSpans> spansByLibrary = new Dictionary<string, ScopeSpans>();
-        ResourceSpans resourceSpans = new ResourceSpans
-        {
-            Resource = processResource,
-        };
-        request.ResourceSpans.Add(resourceSpans);
+        var spansByLibrary = Interlocked.Exchange(ref spansByLibraryCached, null) ?? new Dictionary<string, ScopeSpans>();
```

Comment:
@paulomorgado - We could define this field as `[ThreadStatic]`

```csharp
[ThreadStatic]
private static Dictionary<string, ScopeSpans>? spansByLibrary;
....
....
...

internal static void AddBatch(..)
{
     spansByLibrary ??= new Dictionary<string, ScopeSpans>();
..
..
..
```

Reply:
@vishweshbankwar,

> @paulomorgado - We could define this field as `[ThreadStatic]`
> 
> ```cs
> [ThreadStatic]
> private static Dictionary<string, ScopeSpans>? spansByLibrary;
> ....
> ....
> ...
> 
> internal static void AddBatch(..)
> {
>      spansByLibrary ??= new Dictionary<string, ScopeSpans>();
> ..
> ..
> ..
> ```

How many concurrent threads do you expect to be concurrently requesting these dictionaries?

Reply:
There is a dedicated Exporter thread and that should be the only one invoking this code path.

Reply:
We can benchmark it, but accessing thread local storage might be worse than interlocked operations.

Reply:
> How many concurrent threads do you expect to be concurrently requesting these dictionaries?

it depends - With single instance of exporter configured as BatchExportProcessor. There would be only one thread executing this. With multiple instances, there could be multiple threads executing this.

Reply:
could we just use a normal instance variable? no need of static or threadlocals.

Reply:
we could do that. That would require some refactoring and I would not recommend that effort as the implementation will change anyways with https://github.com/open-telemetry/opentelemetry-dotnet/issues/5450.

Reply:
> we could do that. That would require some refactoring and I would not recommend that effort as the implementation will change anyways with #5450.

is the concern that #5450 will change OTLP Exporter internals, so its better to not make any changes there for now? 
If that is the case, then lets close this PR itself as we should not be proceeding with any approaches being discussed, until after the rewriting is complete.

Reply:
> There is a dedicated Exporter thread and that should be the only one invoking this code path.

Isn't this code path called for `SimpleExportProcessor` as well?

Reply:
> is the concern that https://github.com/open-telemetry/opentelemetry-dotnet/issues/5450 will change OTLP Exporter internals, so its better to not make any changes there for now

Yes

> If that is the case, then lets close this PR itself as we should not be proceeding with any approaches being discussed, until after the rewriting is complete

Agree. I was open to do the small change until #5450 completes. Already have #5731 now, we can close this.

Reply:
> @utpilla 

> Isn't this [code path](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry.Exporter.OpenTelemetryProtocol/OtlpTraceExporter.cs#L65) called for SimpleExportProcessor as well?

Yes - that is also called within the lock right?

Reply:
Yup. I was mainly concerned with `HashMap` being `static`. The issue would mainly be with multiple instances of exporter:
- Multiple SimpleExportProcessor instances
- Multiple BatchExportProcessor instances
- Or a combination of both